File: Backend/controllers/authController.js
const User = require('../models/userModel');
const { validateLogin, validateUser } = require('../utils/validation');
const jwt = require('jsonwebtoken');
const dotenv = require('dotenv');
dotenv.config();

const AuthController = {
  register: async (req, res) => {
    try {
      const { username, email, password } = req.body;
      const errors = validateUser({ username, email, password });
      if (errors.length) return res.status(400).json({ success: false, message: errors.join(', ') });

      const existing = await User.findByUsername(username);
      if (existing) return res.status(400).json({ success: false, message: 'Username sudah digunakan' });

      const id = await User.create({ username, email, password, role: 'user' });
      const user = await User.findById(id);

      const token = jwt.sign(
        { id: user.id, username: user.username, email: user.email, role: user.role },
        process.env.JWT_SECRET,
        { expiresIn: process.env.JWT_EXPIRES_IN }
      );

      res.status(201).json({
        success: true,
        data: {
          token,
          user: { id: user.id, username: user.username, email: user.email, role: user.role }
        }
      });
    } catch (error) {
      console.error(error);
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  login: async (req, res) => {
    try {
      const { username, password } = req.body;
      const errors = validateLogin({ username, password });
      if (errors.length) return res.status(400).json({ success: false, message: errors.join(', ') });

      const user = await User.findByUsername(username);
      if (!user) return res.status(401).json({ success: false, message: 'Username atau password salah' });

      const valid = await User.comparePassword(password, user.password);
      if (!valid) return res.status(401).json({ success: false, message: 'Username atau password salah' });

      const token = jwt.sign(
        { id: user.id, username: user.username, email: user.email, role: user.role },
        process.env.JWT_SECRET,
        { expiresIn: process.env.JWT_EXPIRES_IN }
      );

      res.json({
        success: true,
        data: {
          token,
          user: { id: user.id, username: user.username, email: user.email, role: user.role }
        }
      });
    } catch (error) {
      console.error(error);
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  getProfile: async (req, res) => {
    try {
      const user = await User.findById(req.user.id);
      if (!user) return res.status(404).json({ success: false, message: 'User tidak ditemukan' });
      res.json({ success: true, data: user });
    } catch (error) {
      res.status(500).json({ success: false, message: 'Server error' });
    }
  },

  logout: (req, res) => {
    res.json({ success: true, message: 'Logout berhasil' });
  }
};

module.exports = AuthController;

File: backend_python/controllers/auth_controller.py
from flask import g, request

from middlewares.auth import sign_token
from models.user_model import User
from utils.helpers import fail, ok
from utils.validation import validate_login, validate_user


class AuthController:
    @staticmethod
    def register():
        body = request.get_json(silent=True) or {}
        username, email, password = body.get("username"), body.get("email"), body.get("password")
        errors = validate_user({"username": username, "email": email, "password": password})
        if errors:
            return fail(", ".join(errors), 400)

        existing = User.find_by_username(username)
        if existing:
            return fail("Username sudah digunakan", 400)

        user_id = User.create({"username": username, "email": email, "password": password, "role": "user"})
        user = User.find_by_id(user_id)
        token = sign_token(user)
        data = {
            "token": token,
            "user": {"id": user["id"], "username": user["username"], "email": user["email"], "role": user["role"]},
        }
        return ok({"success": True, "data": data}, 201)

    @staticmethod
    def login():
        body = request.get_json(silent=True) or {}
        username, password = body.get("username"), body.get("password")
        errors = validate_login({"username": username, "password": password})
        if errors:
            return fail(", ".join(errors), 400)

        user = User.find_by_username(username)
        if not user:
            return fail("Username atau password salah", 401)

        valid = User.compare_password(password, user["password"])
        if not valid:
            return fail("Username atau password salah", 401)

        token = sign_token(user)
        data = {
            "token": token,
            "user": {"id": user["id"], "username": user["username"], "email": user["email"], "role": user["role"]},
        }
        return ok({"success": True, "data": data})

    @staticmethod
    def get_profile():
        user = User.find_by_id(g.user["id"])
        if not user:
            return fail("User tidak ditemukan", 404)
        return ok({"data": user})

    @staticmethod
    def logout():
        from flask import jsonify

        return jsonify({"success": True, "message": "Logout berhasil"})
